src/core/localize.rkt

```diff
@@ -164,52 +177,84 @@
     (define exacts (list->vector (apply subexprs-fn pt)))
     (define actuals (apply actual-value-fn pt))
 
-    (for ([expr (in-list exprs-list)]
+    (for ([spec (in-vector spec-vec)]
+          [expr (in-list exprs-list)]
           [root (in-vector roots)]
           [exact (in-vector exacts)]
           [actual (in-vector actuals)]
           [expr-idx (in-naturals)])
-      (define err
+      (define true-err
+        (if true-err? ;; ??? Whats the default values for true error literal, variable approx and if?
+            (match (vector-ref nodes root)
+              [(? literal?) 0]
+              [(? variable?) 0]
+              [(approx aprx-spec impl) exact] ;; TODO not sure what to do here.
+              [`(if ,c ,ift ,iff) 0]
+              [(list f args-roots ...)
+               ;; Find the index of the variables we need to substitute.
+               (match exact
+                 [`+nan.0 `+nan.0]
+                 [`-nan.0 `-nan.0]
+                 [`+inf.0 `+inf.0]
+                 [`-inf.0 `-inf.0]
+                 [value
+                  ; __exact double underscore to avoid conflicts with user provided
+                  ; variables. Could use name mangling long term.
+                  (define modifed-vars (append all-vars `(__exact)))
+                  (define true-error-expr (list `(- ,spec __exact)))
+                  (define diffMachine
+                    (rival-compile true-error-expr modifed-vars (list flonum-discretization)))
```

Comment:
If this is UI-only maybe it's fine but this way of doing it is slow, because you're compiling a separate machine for every single subexpression and every single input. It would be more efficient to do this in two passes: first compute the exact value for each subexpression for each input; then compile a second machine with one input and output per subexpression and feed in each input + each subexpression's exact value for that input.

Reply:
Ok i'm a little confused on this. Do you mean something like?

```
(define exact-var-name '__exact)
  (define extended
    (for/list ([ctx (in-list ctx-list)])
      (context-extend ctx exact-var-name (context-repr ctx))))
  (define compare-specs
    (for/list ([spec (in-list spec-list)])
      `(- ,spec ,exact-var-name)))
  (define compare-fn (eval-progs-real compare-specs extended))
```

Reply:
Look, let's say you have three specs: `1`, `x`, and `1 + x`. Then you have three actual values for them: `1.0`, `1e20`, and `1e20`. Then you want to evaluate three new expressions: `1 - 1.0`, `x - 1e20`, and `(1 + x) - 1e20`.

Now let's generalize from this example. Let's say we have expressions `expr[0]`, `expr[1]`, ..., `expr[n]`. We'll have an actual value for each of them: `actuals[0]`, `actuals[1]`, ..., `actuals[n]`. Now for each one you'll want to evaluate `(- expr[i] actuals[i])`.

I think you can just stick the actual value right in the expression, but if for some reason you want to use a variable you can do that too. You would need one variable for each expression: `var[0]`, var[1]`, ..., `var[n]`. And then you'd need to form the expressions `(- expr[i] var[i])`. You could `eval-progs-real` to get a function `f`, and you'd want to call `f(actuals[0], actuals[1], ..., actuals[n], pt ...)` to get the differences.